src/Microsoft.Diagnostics.Runtime/ClrValue.cs

```diff
@@ -0,0 +1,1077 @@
+                throw new ArgumentException($"Type '{type.Name}' does not contain a field named '{fieldName}'");
+
+            if (!field.IsObjectReference)
+                throw new ArgumentException($"Field '{type.Name}.{fieldName}' is not an object reference.");
+
+            ClrHeap heap = Type.Heap;
+
+            ulong addr = ClrRuntime.IsObjectReference(ElementType) ? Object : Address;
+            addr = field.GetAddress(addr, Interior);
+            ulong obj;
+
+            if (!heap.ReadPointer(addr, out obj))
+                throw new MemoryReadException(addr);
+            
+            return new ClrObject(obj, heap.GetObjectType(obj));
+        }
+
+        /// <summary>
+        /// Retrieves a field from this value.
+        /// </summary>
+        /// <param name="name">The name of the field.</param>
+        /// <returns>A ClrValue representing this field.</returns>
+        public virtual ClrValue GetField(string name)
+        {
+            ClrElementType el = ElementType;
+            if (ClrRuntime.IsPrimitive(el))
+            {
+                // Primitives only have one field, named m_value.
+                if (name != "m_value")
+                    throw new ArgumentException(string.Format("Field '{0}' does not exist in type '{1}'.", name, Type.Name));
+
+                // Getting m_value is the same as this ClrValue...
+                return this;
+            }
+
+            if (ClrRuntime.IsObjectReference(el) || !Interior)
+                return AsObject().GetField(name);
+
+            Debug.Assert(ClrRuntime.IsValueClass(el));
+
+            ulong address = Address;
+            if (address == 0)
+                throw new NullReferenceException();
+
+            ClrType type = Type;
+            ClrInstanceField field = type.GetFieldByName(name);
+            if (field == null)
+                throw new ArgumentException(string.Format("Field '{0}' does not exist in type '{1}'.", name, Type.Name));
+
+            ulong result = field.GetAddress(address, Interior);
+            return new ClrValueImpl(_runtime, result, field);
+        }
+
+        /// <summary>
+        /// Gets a boolean field from the object.  Note that the type must match exactly, as this method
+        /// will not do type coercion.  This method will throw an ArgumentException if no field matches
+        /// the given name.  It will throw a NullReferenceException if the target object is null (that is,
+        /// if (IsNull returns true).  It will throw an InvalidOperationException if the field is not
+        /// of the correct type.  Lastly, it will throw a MemoryReadException if there was an error reading
+        /// the value of this field out of the data target.
+        /// </summary>
+        /// <param name="fieldName">The name of the field to get the value for.</param>
+        /// <returns>The value of the given field.</returns>
+        public bool GetBoolean(string fieldName)
+        {
```

Comment:
Ah so the problem is generally the last step.  Let's say you already have the raw address of where to read from and you know the type T of the variable you are trying to convert to.  There's not a great way to way to convert a byte[] in a generic Foo<T> to a T.

Like let's say you have a byte[] of the value already read, and you have to convert it to T.  The general way to convert a byte[] to (let's say) a boolean is to call BitConverter.ToBoolean().  As far as I can tell you will have to either convert that boolean to an object and convert it on the way out, or do something funny with the dynamic keyword.

Either of which aren't the end of the world, just slower.  I assume that BitConverter is the class I should model behavior after, and thus we have "GetBoolean, GetInt16, etc).